File: so/fork_c/basico.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/wait.h>

void tarefa(int segundos) {
    sleep(segundos);
}

void filho(int numero, int segundos) {
    printf("Filho %d (PID %d) iniciado\n", numero, getpid());
    tarefa(segundos);
    printf("Filho %d (PID %d) finalizado após %d segundos\n", numero, getpid(), segundos);
    exit(0);  // Finaliza o processo filho com sucesso
}

int main() {
    printf("Pai (PID %d) iniciando...\n", getpid());

    int tempos[] = {3, 2, 4};
    pid_t pids[3];

    for (int i = 0; i < 3; i++) {
        pid_t pid = fork();
        if (pid < 0) {
            perror("Erro ao criar processo filho");
            exit(1);
            } else if (pid == 0) {
            // Código do filho
            filho(i + 1, tempos[i]);
            } else {
            // Código do pai
            pids[i] = pid;
            printf("Pai criou Filho %d com PID %d\n", i + 1, pid);

            // Pai espera os filhos terminarem
            for (int i = 0; i < 3; i++) {
                int status;
                pid_t terminado = waitpid(pids[i], &status, 0);
                if (terminado > 0) {
                    printf("Pai detectou que Filho %d terminou (PID %d)\n", i + 1, terminado);
                }
            }
            printf("Pai finalizado. Todos os filhos terminaram.\n");
        }

    }   
    
    return 0;
}

File: so/fork_python/basico.py
#!/usr/bin/env python3

import os
import time
import sys

def tarefa(segundos):
    time.sleep(segundos)

def filho(numero, segundos):
    print(f"Filho {numero} (PID {os.getpid()}) iniciado")
    tarefa(segundos)
    print(f"Filho {numero} (PID {os.getpid()}) finalizado após {segundos} segundos")
    os._exit(0)  # termina o processo filho

def main():
    print(f"Pai (PID {os.getpid()}) iniciando...")

    tempos = [3, 2, 4]
    pids = []

    for i, tempo in enumerate(tempos, 1):
        pid = os.fork()
        if pid == 0:
            # Código executado no filho
            filho(i, tempo)
        else:
            # Código executado no pai
            pids.append((i, pid))

    # Pai espera os filhos terminarem
    for i, pid in pids:
        terminado, _ = os.waitpid(pid, 0)
        print(f"Pai detectou que Filho {i} terminou (PID {terminado})")

    print("Pai finalizado. Todos os filhos terminaram.")

if __name__ == "__main__":
    if os.name != "posix":
        print("Este script requer um sistema Unix-like (Linux, macOS).", file=sys.stderr)
        sys.exit(1)
    main()